File: V0.2/linkguardMB/LinkGuardHQ/LinkGuardHQ/HQBluetoothManager.swift
import Foundation
import CoreBluetooth
import Combine

// MARK: - HQ BLE 常數（對應 HQ LoRa 韌體 UUID，與 Field 端不同）

struct HQLoRaBLE {
    static let serviceUUID      = CBUUID(string: "4FAFC201-1FB5-459E-8FCC-C5C9C331915B")
    static let statusUUID       = CBUUID(string: "BEB5483E-36E1-4688-B7F5-EA07361B26B8")
    static let commandUUID      = CBUUID(string: "BEB5483E-36E1-4688-B7F5-EA07361B26BB")
    static let loraCommandUUID  = CBUUID(string: "BEB5483E-36E1-4688-B7F5-EA07361B26BC")
}

// MARK: - HQ LoRa 韌體狀態

struct HQLoRaStatus: Decodable {
    let id: String
    let freq: Double
    let bat: Int
    let vbat: Double
    let lvl: Int
    let pair: String
    let uptime: Int
    let hqNodes: [HQLoRaNode]
    let cmdCount: Int
}

struct HQLoRaNode: Decodable, Identifiable {
    let id: String
    let bat: Int
    let rssi: Double
    let snr: Double
    let online: Bool
}

struct HQLoRaReceivedCommand: Decodable {
    let cmd_id: String
    let type: String
    let pri: Int
    let title: String
    let detail: String
    let sender: String
}

// MARK: - 已發現 BLE 裝置

struct HQDiscoveredDevice: Identifiable {
    let id: UUID
    let name: String
    let rssi: Int
    let peripheral: CBPeripheral
}

// MARK: - HQ LoRa 藍牙管理器

class HQBluetoothManager: NSObject, ObservableObject {
    @Published var isScanning = false
    @Published var isConnected = false
    @Published var isPoweredOn = false
    @Published var discoveredDevices: [HQDiscoveredDevice] = []
    @Published var connectedDeviceName: String?
    @Published var loraStatus: HQLoRaStatus?

    private var centralManager: CBCentralManager!
    private var connectedPeripheral: CBPeripheral?
    private var commandCharacteristic: CBCharacteristic?
    private var statusBuffer = Data()
    private var statusBufferTimestamp = Date()
    private var reconnectTimer: Timer?
    private var reconnectDelay: TimeInterval = 2

    var onStatusUpdate: ((HQLoRaStatus) -> Void)?
    var onLoRaCommand: ((HQLoRaReceivedCommand) -> Void)?

    override init() {
        super.init()
        centralManager = CBCentralManager(delegate: self, queue: nil)
    }

    // MARK: - 掃描

    func startScanning() {
        guard centralManager.state == .poweredOn else { return }
        isScanning = true
        discoveredDevices.removeAll()
        centralManager.scanForPeripherals(
            withServices: [HQLoRaBLE.serviceUUID],
            options: [CBCentralManagerScanOptionAllowDuplicatesKey: false]
        )
    }

    func stopScanning() {
        centralManager.stopScan()
        isScanning = false
    }

    // MARK: - 連線

    func connect(to device: HQDiscoveredDevice) {
        stopScanning()
        if let old = connectedPeripheral {
            centralManager.cancelPeripheralConnection(old)
        }
        connectedPeripheral = device.peripheral
        centralManager.connect(device.peripheral, options: nil)
    }

    func disconnect() {
        reconnectTimer?.invalidate()
        reconnectDelay = 2
        guard let p = connectedPeripheral else { return }
        connectedPeripheral = nil
        connectedDeviceName = nil
        centralManager.cancelPeripheralConnection(p)
    }

    // MARK: - 發送指令

    func sendLevelChange(_ level: Int) {
        sendBLE("setlvl:\(level)")
    }

    func sendPairChange(_ pair: String) {
        sendBLE("setpair:\(pair)")
    }

    func sendCommand(cmdId: String, type: String, priority: Int, title: String, detail: String) {
        sendBLE("cmd:\(cmdId):\(type):\(priority):\(title):\(detail)")
    }

    func sendBroadcast(msgType: String, payload: String) {
        sendBLE("broadcast:\(msgType):\(payload)")
    }

    private func sendBLE(_ text: String) {
        guard let char = commandCharacteristic,
              let p = connectedPeripheral,
              let data = text.data(using: .utf8)
        else { return }
        p.writeValue(data, for: char, type: .withResponse)
    }
}

// MARK: - CBCentralManagerDelegate

extension HQBluetoothManager: CBCentralManagerDelegate {

    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        isPoweredOn = central.state == .poweredOn
        if central.state != .poweredOn {
            isConnected = false
            isScanning = false
        }
    }

    func centralManager(_ central: CBCentralManager,
                         didDiscover peripheral: CBPeripheral,
                         advertisementData: [String: Any],
                         rssi RSSI: NSNumber) {
        guard !discoveredDevices.contains(where: { $0.id == peripheral.identifier }) else { return }
        let name = peripheral.name ?? "HQ LoRa"
        discoveredDevices.append(
            HQDiscoveredDevice(id: peripheral.identifier, name: name, rssi: RSSI.intValue, peripheral: peripheral)
        )
    }

    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        isConnected = true
        reconnectTimer?.invalidate()
        reconnectDelay = 2
        connectedDeviceName = peripheral.name ?? "HQ LoRa"
        peripheral.delegate = self
        peripheral.discoverServices([HQLoRaBLE.serviceUUID])
    }

    func centralManager(_ central: CBCentralManager,
                         didDisconnectPeripheral peripheral: CBPeripheral,
                         error: Error?) {
        isConnected = false
        commandCharacteristic = nil
        loraStatus = nil
        statusBuffer.removeAll()
        // 自動重連（指數退避：2s → 4s → 8s … 最多 60s）
        if connectedPeripheral != nil {
            print("[HQ-BLE] Disconnected, will retry in \(reconnectDelay)s...")
            reconnectTimer?.invalidate()
            reconnectTimer = Timer.scheduledTimer(withTimeInterval: reconnectDelay, repeats: false) { [weak self] _ in
                guard let self, let p = self.connectedPeripheral, self.centralManager.state == .poweredOn else { return }
                print("[HQ-BLE] Attempting reconnect...")
                self.centralManager.connect(p, options: nil)
            }
            reconnectDelay = min(reconnectDelay * 2, 60)
        }
    }

    func centralManager(_ central: CBCentralManager,
                         didFailToConnect peripheral: CBPeripheral,
                         error: Error?) {
        isConnected = false
        connectedPeripheral = nil
    }
}

// MARK: - CBPeripheralDelegate

extension HQBluetoothManager: CBPeripheralDelegate {

    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        guard error == nil, let services = peripheral.services else { return }
        for service in services {
            peripheral.discoverCharacteristics([
                HQLoRaBLE.statusUUID,
                HQLoRaBLE.commandUUID,
                HQLoRaBLE.loraCommandUUID
            ], for: service)
        }
    }

    func peripheral(_ peripheral: CBPeripheral,
                     didDiscoverCharacteristicsFor service: CBService,
                     error: Error?) {
        guard let chars = service.characteristics else { return }
        for char in chars {
            switch char.uuid {
            case HQLoRaBLE.statusUUID:
                peripheral.setNotifyValue(true, for: char)
            case HQLoRaBLE.commandUUID:
                commandCharacteristic = char
            case HQLoRaBLE.loraCommandUUID:
                peripheral.setNotifyValue(true, for: char)
            default:
                break
            }
        }
    }

    func peripheral(_ peripheral: CBPeripheral,
                     didUpdateValueFor characteristic: CBCharacteristic,
                     error: Error?) {
        guard error == nil, let data = characteristic.value, !data.isEmpty else { return }

        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            switch characteristic.uuid {
            case HQLoRaBLE.statusUUID:
                if let firstByte = data.first, firstByte == UInt8(ascii: "{") {
                    self.statusBuffer = data
                    self.statusBufferTimestamp = Date()
                } else {
                    self.statusBuffer.append(data)
                }
                if let status = try? JSONDecoder().decode(HQLoRaStatus.self, from: self.statusBuffer) {
                    self.statusBuffer.removeAll()
                    self.loraStatus = status
                    self.onStatusUpdate?(status)
                } else if self.statusBuffer.count > 4096 ||
                          Date().timeIntervalSince(self.statusBufferTimestamp) > 3 {
                    self.statusBuffer.removeAll()
                }
            case HQLoRaBLE.loraCommandUUID:
                if let cmd = try? JSONDecoder().decode(HQLoRaReceivedCommand.self, from: data) {
                    self.onLoRaCommand?(cmd)
                }
            default:
                break
            }
        }
    }

    func peripheral(_ peripheral: CBPeripheral,
                     didUpdateNotificationStateFor characteristic: CBCharacteristic,
                     error: Error?) {
        if let error = error {
            print("[HQ-BLE] Notify subscribe failed: \(error)")
        }
    }

    func peripheral(_ peripheral: CBPeripheral, didModifyServices invalidatedServices: [CBService]) {
        peripheral.discoverServices([HQLoRaBLE.serviceUUID])
    }

    func peripheral(_ peripheral: CBPeripheral,
                     didWriteValueFor characteristic: CBCharacteristic,
                     error: Error?) {
        if let error = error {
            print("[HQ-BLE] Write failed: \(error)")
        }
    }
}

File: V0.2/linkguardMB/LinkGuardHQ-Android/app/src/main/java/com/linkguard/hq/ble/HQBluetoothManager.kt
package com.linkguard.hq.ble

import android.annotation.SuppressLint
import android.bluetooth.*
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.os.ParcelUuid
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import org.json.JSONObject
import java.util.UUID

// BLE UUIDs matching HQ LoRa firmware
object HQLoRaBLE {
    val SERVICE_UUID: UUID      = UUID.fromString("4FAFC201-1FB5-459E-8FCC-C5C9C331915B")
    val STATUS_UUID: UUID       = UUID.fromString("BEB5483E-36E1-4688-B7F5-EA07361B26B8")
    val COMMAND_UUID: UUID      = UUID.fromString("BEB5483E-36E1-4688-B7F5-EA07361B26BB")
    val LORA_COMMAND_UUID: UUID = UUID.fromString("BEB5483E-36E1-4688-B7F5-EA07361B26BC")
    val CCCD_UUID: UUID         = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
}

data class HQLoRaNode(
    val id: String,
    val bat: Int,
    val rssi: Double,
    val snr: Double,
    val online: Boolean
)

data class HQLoRaStatus(
    val id: String,
    val freq: Double,
    val bat: Int,
    val vbat: Double,
    val lvl: Int,
    val pair: String,
    val uptime: Int,
    val hqNodes: List<HQLoRaNode>,
    val cmdCount: Int
)

data class HQDiscoveredDevice(
    val address: String,
    val name: String,
    val rssi: Int
)

@SuppressLint("MissingPermission")
class HQBluetoothManager(private val context: Context) {

    private val _isScanning = MutableStateFlow(false)
    val isScanning: StateFlow<Boolean> = _isScanning

    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected

    private val _isPoweredOn = MutableStateFlow(false)
    val isPoweredOn: StateFlow<Boolean> = _isPoweredOn

    private val _discoveredDevices = MutableStateFlow<List<HQDiscoveredDevice>>(emptyList())
    val discoveredDevices: StateFlow<List<HQDiscoveredDevice>> = _discoveredDevices

    private val _connectedDeviceName = MutableStateFlow<String?>(null)
    val connectedDeviceName: StateFlow<String?> = _connectedDeviceName

    private val _loraStatus = MutableStateFlow<HQLoRaStatus?>(null)
    val loraStatus: StateFlow<HQLoRaStatus?> = _loraStatus

    var onStatusUpdate: ((HQLoRaStatus) -> Unit)? = null
    var onLoRaCommand: ((String) -> Unit)? = null

    private val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
    private val bluetoothAdapter: BluetoothAdapter? = bluetoothManager?.adapter
    private val scanner = bluetoothAdapter?.bluetoothLeScanner

    private var connectedGatt: BluetoothGatt? = null
    private var commandCharacteristic: BluetoothGattCharacteristic? = null
    private var targetDevice: BluetoothDevice? = null

    private var statusBuffer = StringBuilder()
    private var reconnectScope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    private var reconnectDelay = 2000L
    private var reconnectJob: Job? = null

    init {
        _isPoweredOn.value = bluetoothAdapter?.isEnabled == true
    }

    // Scanning

    fun startScanning() {
        if (scanner == null || bluetoothAdapter?.isEnabled != true) return
        _isScanning.value = true
        _discoveredDevices.value = emptyList()

        val filter = ScanFilter.Builder()
            .setServiceUuid(ParcelUuid(HQLoRaBLE.SERVICE_UUID))
            .build()
        val settings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .build()

        scanner.startScan(listOf(filter), settings, scanCallback)
    }

    fun stopScanning() {
        try { scanner?.stopScan(scanCallback) } catch (_: Exception) {}
        _isScanning.value = false
    }

    private val scanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            val device = result.device
            val address = device.address
            val current = _discoveredDevices.value
            if (current.none { it.address == address }) {
                _discoveredDevices.value = current + HQDiscoveredDevice(
                    address = address,
                    name = device.name ?: "HQ LoRa",
                    rssi = result.rssi
                )
            }
        }
    }

    // Connection

    fun connect(device: HQDiscoveredDevice) {
        stopScanning()
        connectedGatt?.close()
        val btDevice = bluetoothAdapter?.getRemoteDevice(device.address) ?: return
        targetDevice = btDevice
        reconnectGatt(btDevice)
    }

    private fun reconnectGatt(btDevice: BluetoothDevice) {
        btDevice.connectGatt(context, false, gattCallback, BluetoothDevice.TRANSPORT_LE)
    }

    fun disconnect() {
        reconnectJob?.cancel()
        reconnectDelay = 2000L
        targetDevice = null
        connectedGatt?.disconnect()
        connectedGatt?.close()
        connectedGatt = null
        _isConnected.value = false
        _connectedDeviceName.value = null
        _loraStatus.value = null
        commandCharacteristic = null
    }

    // Send commands

    fun sendLevelChange(level: Int) = sendBLE("setlvl:$level")
    fun sendPairChange(pair: String) = sendBLE("setpair:$pair")
    fun sendCommand(cmdId: String, type: String, priority: Int, title: String, detail: String) =
        sendBLE("cmd:$cmdId:$type:$priority:$title:$detail")
    fun sendBroadcast(msgType: String, payload: String) = sendBLE("broadcast:$msgType:$payload")

    private fun sendBLE(text: String) {
        val char = commandCharacteristic ?: return
        val gatt = connectedGatt ?: return
        char.value = text.toByteArray(Charsets.UTF_8)
        gatt.writeCharacteristic(char)
    }

    fun destroy() {
        reconnectJob?.cancel()
        reconnectScope.cancel()
        disconnect()
    }

    // GATT callback

    private val gattCallback: BluetoothGattCallback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            when (newState) {
                BluetoothProfile.STATE_CONNECTED -> {
                    connectedGatt = gatt
                    _isConnected.value = true
                    _connectedDeviceName.value = gatt.device.name ?: "HQ LoRa"
                    reconnectJob?.cancel()
                    reconnectDelay = 2000L
                    gatt.discoverServices()
                }
                BluetoothProfile.STATE_DISCONNECTED -> {
                    _isConnected.value = false
                    commandCharacteristic = null
                    _loraStatus.value = null
                    statusBuffer.clear()
                    // Auto-reconnect with exponential backoff (2s → 60s)
                    if (targetDevice != null) {
                        reconnectJob?.cancel()
                        reconnectJob = reconnectScope.launch {
                            delay(reconnectDelay)
                            reconnectDelay = (reconnectDelay * 2).coerceAtMost(60000L)
                            targetDevice?.let { reconnectGatt(it) }
                        }
                    }
                }
            }
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            if (status != BluetoothGatt.GATT_SUCCESS) return
            val service = gatt.getService(HQLoRaBLE.SERVICE_UUID) ?: return

            service.getCharacteristic(HQLoRaBLE.STATUS_UUID)?.let { char ->
                gatt.setCharacteristicNotification(char, true)
                char.getDescriptor(HQLoRaBLE.CCCD_UUID)?.let { desc ->
                    desc.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                    gatt.writeDescriptor(desc)
                }
            }

            service.getCharacteristic(HQLoRaBLE.COMMAND_UUID)?.let { char ->
                commandCharacteristic = char
            }

            service.getCharacteristic(HQLoRaBLE.LORA_COMMAND_UUID)?.let { char ->
                gatt.setCharacteristicNotification(char, true)
                char.getDescriptor(HQLoRaBLE.CCCD_UUID)?.let { desc ->
                    desc.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                    gatt.writeDescriptor(desc)
                }
            }
        }

        override fun onCharacteristicChanged(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
            val data = characteristic.value ?: return
            val text = String(data, Charsets.UTF_8)

            when (characteristic.uuid) {
                HQLoRaBLE.STATUS_UUID -> {
                    if (text.startsWith("{")) {
                        statusBuffer.clear()
                        statusBuffer.append(text)
                    } else {
                        statusBuffer.append(text)
                    }
                    // Try to parse complete JSON
                    val json = statusBuffer.toString()
                    if (json.contains("}")) {
                        try {
                            val parsed = parseLoRaStatus(json)
                            _loraStatus.value = parsed
                            onStatusUpdate?.invoke(parsed)
                            statusBuffer.clear()
                        } catch (_: Exception) {
                            // Incomplete JSON, wait for more data
                        }
                    }
                }
                HQLoRaBLE.LORA_COMMAND_UUID -> {
                    onLoRaCommand?.invoke(text)
                }
            }
        }
    }

    private fun parseLoRaStatus(json: String): HQLoRaStatus {
        val obj = JSONObject(json)
        val nodesArray = obj.optJSONArray("hqNodes") ?: org.json.JSONArray()
        val nodes = mutableListOf<HQLoRaNode>()
        for (i in 0 until nodesArray.length()) {
            val n = nodesArray.getJSONObject(i)
            nodes.add(HQLoRaNode(
                id = n.optString("id", ""),
                bat = n.optInt("bat", 0),
                rssi = n.optDouble("rssi", 0.0),
                snr = n.optDouble("snr", 0.0),
                online = n.optBoolean("online", false)
            ))
        }
        return HQLoRaStatus(
            id = obj.optString("id", ""),
            freq = obj.optDouble("freq", 0.0),
            bat = obj.optInt("bat", 0),
            vbat = obj.optDouble("vbat", 0.0),
            lvl = obj.optInt("lvl", 0),
            pair = obj.optString("pair", ""),
            uptime = obj.optInt("uptime", 0),
            hqNodes = nodes,
            cmdCount = obj.optInt("cmdCount", 0)
        )
    }
}
